mitmproxy/addons/session.py

```diff
@@ -67,7 +111,278 @@ def is_session_db(path):
+            await asyncio.sleep(self._flush_period)
+            tof = []
+            to_dump = min(self._flush_rate, len(self._hot_store))
+            for _ in range(to_dump):
+                tof.append(self._hot_store.popitem(last=False)[1])
+            self.db_store.store_flows(tof)
+
+    async def _tweaker(self):
+        while True:
+            await asyncio.sleep(self._tweak_period)
+            if len(self._hot_store) >= 3 * self._flush_rate:
+                self._flush_period *= 0.9
+                self._flush_rate *= 1.1
+            elif len(self._hot_store) < self._flush_rate:
+                self._flush_period *= 1.1
+                self._flush_rate *= 0.9
+
+    def load_view(self):
+        ids = [fid for _, fid in self._view]
+        flows = self.load_storage(ids)
+        return sorted(flows, key=lambda f: self._generate_order(f))
+
+    def load_storage(self, ids=None):
+        flows = []
+        ids_from_store = []
+        if ids is not None:
+            for fid in ids:
+                # A same flow could be at the same time in hot and db storage. We want the most updated version.
+                if fid in self._hot_store:
+                    flows.append(self._hot_store[fid])
+                elif fid in self.db_store:
+                    ids_from_store.append(fid)
+                else:
+                    flows.append(None)
+            flows += self.db_store.retrieve_flows(ids_from_store)
+        else:
+            for flow in self._hot_store.values():
+                flows.append(flow)
+            for flow in self.db_store.retrieve_flows():
+                if flow.id not in self._hot_store:
+                    flows.append(flow)
+        return flows
+
+    def clear_storage(self):
+        self.db_store.clear()
+        self._hot_store.clear()
+        self._view = []
+
+    def store_count(self):
+        ln = 0
+        for fid in self._hot_store.keys():
+            if fid not in self.db_store:
+                ln += 1
+        return ln + len(self.db_store)
+
+    def _generate_order(self, f: http.HTTPFlow) -> typing.Optional[typing.Union[str, int, float]]:
+        o = self.order
+        if o == "time":
+            return f.request.timestamp_start or 0
+        if o == "method":
+            return f.request.method
+        if o == "url":
+            return f.request.url
+        if o == "size":
+            s = 0
+            if f.request.raw_content:
+                s += len(f.request.raw_content)
+            if f.response and f.response.raw_content:
+                s += len(f.response.raw_content)
+            return s
+        return None
+
+    def set_order(self, order: str) -> None:
+        if order not in orders:
+            raise CommandError(
+                "Unknown flow order: %s" % order
+            )
+        if order != self.order:
+            self.order = order
+            newview = [
+                (self._generate_order(f), f.id) for f in self.load_view()
```

Comment:
This seems to reload all flows from disk. Changing the order of flows on the fly is going to be a very important operation, and needs to be very snappy. I'd rather have _generate_order generate a dict with all the order keys, so we can do this once. Then we can make changing the order an operation that just changes the sort key and re-sorts.